### PR TITLE
Fix deprecated strtolower on order received page

### LINKED ISSUE #8232: [GlobalStep] PHP deprecated error "strtolower(): Passing null to parameter #1 ($string) of type string" triggered on "Order received" page.
### Bug Description:
PHP deprecated error "strtolower(): Passing null to parameter ($string) of type string" triggered on "Order received" page.

### Environment:
Woocommerce Version: WooCommerce 8.6.0
WooCommerce Payments: 7.3.0-test-1
WordPress version: v6.4.3

PC:
Windows 10,
Chrome(Version 121.0.6167.185)
Firefox (Version 122.0.1)

### Steps To Reproduce:
1. Create any test site.
2. Install and activate all the required plugins (i.e Query monitor).
3. Install the WooCommerce Payments.
4. Complete WooCommerce Payments KYC flow.
5. Go to the store page and add a product to your cart.
6. Go to Checkout page and click place order.
7. User should be redirected on "Order received" page.
8. Observe that, PHP deprecated error "strtolower(): Passing null to parameter ($string) of type string" triggered on "Order received" page.

#### Stack Trace:
[strtolower() Passing null to parameter #1 ($string) of type string is deprecated.txt](https://github.com/Automattic/woocommerce-payments/files/14341271/strtolower.Passing.null.to.parameter.1.string.of.type.string.is.deprecated.txt)

### Actual Result:
PHP deprecated error "strtolower(): Passing null to parameter ($string) of type string" triggered on "Order received" page.

### Expected Result:
Order received page should be displayed without any error.

### Screenshot:

*Isolating the problem (mark completed items with an [x]):*
- [ ] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [ ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

<details>

`
### WordPress Environment ###

WC Version: 8.6.0
REST API Version: ✔ 8.6.0
WC Blocks Version: ✔ 11.8.0-dev
Action Scheduler Version: ✔ 3.7.1
Log Directory Writable: ✔
WP Version: 6.4.3
WP Multisite: –
WP Memory Limit: 512 MB
WP Debug Mode: –
WP Cron: ✔
Language: en_US
External object cache: ✔

### Server Environment ###

Server Info: nginx
PHP Version: 8.1.27
PHP Post Max Size: 2 GB
PHP Time Limit: 1200
PHP Max Input Vars: 6144
cURL Version: 8.4.0
OpenSSL/1.1.1w

SUHOSIN Installed: –
MySQL Version: 10.6.15-MariaDB-log
Max Upload Size: 2 GB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 8.6.0
WC Database Prefix: wp_
Total Database Size: 3.57MB
Database Data Size: 1.69MB
Database Index Size: 1.88MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.06MB + Index: 0.11MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.05MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.05MB + Index: 0.09MB + Engine InnoDB
wp_jetpack_sync_queue: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 0.30MB + Index: 0.06MB + Engine InnoDB
wp_postmeta: Data: 0.11MB + Index: 0.06MB + Engine InnoDB
wp_posts: Data: 0.08MB + Index: 0.06MB + Engine InnoDB
wp_snippets: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.06MB + Index: 0.00MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.06MB + Index: 0.02MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_orders: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_wc_orders_meta: Data: 0.06MB + Index: 0.09MB + Engine InnoDB
wp_wc_order_addresses: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_operational_data: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB

### Post Type Counts ###

attachment: 24
page: 7
post: 1
product: 18
product_variation: 7
shop_order_placehold: 15
wp_navigation: 1
wp_template: 2

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (5) ###

Query Monitor: by John Blackbourn – 3.15.0
Code Snippets: by Code Snippets Pro – 3.6.2
Companion Plugin: by Osk – 1.30
WooPayments: by Automattic – 7.3.0-test-1
WooCommerce: by Automattic – 8.6.0

### Inactive Plugins (1) ###

Akismet Anti-spam: Spam Protection: by Automattic - Anti-spam Team – 5.3.1

### Dropin Plugins (3) ###

advanced-cache.php: advanced-cache.php
db.php: Query Monitor Database Class (Drop-in)
object-cache.php: Memcached

### Settings ###

API Enabled: –
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to Woo.com: –
Enforce Approved Product Download Directories: ✔
HPOS feature screen enabled: ✔
HPOS feature enabled: ✔
Order datastore: Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore
HPOS data sync enabled: –

### WC Pages ###

Shop base: #5 - /shop/
Cart: #6 - /cart/
Checkout: #7 - /checkout/
My account: #8 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Storefront
Version: 4.5.4
Author URL: https://woo.com/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### WooPayments ###

Version: 7.3.0-test-1
Connected to WPCOM: Yes
WPCOM Blog ID: 229644254
Account ID: acct_1OlnP1CAVGMSj8Mh
Payment Gateway: Enabled
Test Mode: Enabled
Enabled APMs: card
WooPay: Disabled
WooPay Incompatible Extensions: No
Apple Pay / Google Pay: Enabled (product,cart,checkout)
Fraud Protection Level: basic
Multi-currency: Enabled
Public Key Encryption: Disabled
Auth and Capture: Enabled
Documents: Disabled
Logging: Enabled

### Admin ###

Enabled Features: activity-panels
analytics
product-block-editor
coupons
core-profiler
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
product-variation-management
product-virtual-downloadable
product-external-affiliate
product-grouped
product-linked
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page

Disabled Features: customize-store
minified-js
new-product-management-experience
product-pre-publish-modal
settings
async-product-editor-category-field

Daily Cron: ✔ Next scheduled: 2024-02-21 07:01:03 +00:00
Options: ✔
Notes: 68
Onboarding: skipped

### Action Scheduler ###

Canceled: 8
Oldest: 2024-02-20 07:16:20 +0000
Newest: 2024-02-20 07:43:06 +0000

Complete: 116
Oldest: 2024-02-20 07:02:23 +0000
Newest: 2024-02-20 07:51:01 +0000

Failed: 2
Oldest: 2024-02-20 07:03:28 +0000
Newest: 2024-02-20 07:13:03 +0000

Pending: 1
Oldest: 2024-02-21 07:02:23 +0000
Newest: 2024-02-21 07:02:23 +0000

### Status report information ###

Generated at: 2024-02-20 07:58:38 +00:00
`

</details>

### PR DESCRIPTION
**Pull request by reykjalin**
Fixes #8232

## Changes proposed in this Pull Request

* Make sure order currency is initialized in the `FrontendCurrencies` class by the time the decimal point formatting function is called.

#### Testing instructions

1. Make sure your store is running on PHP 8.1 or newer.
2. Place an order.
3. Open the Order Received page.
4. Make sure there are no deprecation warnings.
5. Go to **My Account &rarr; Orders** and view one of the orders.
6. Make sure there are no deprecation warnings. 

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `8387` or branch name `fix/8232-deprecated-strtolower-on-order-received-page` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=fix/8232-deprecated-strtolower-on-order-received-page) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 0819154342f223af3df96ddca8f793d86bffb5e5
- Build time: 2024-03-18 19:06:29 UTC

_Note: the build is updated when a new commit is pushed to this PR._

**Comment by reykjalin** [comment]
@jessepearson I'm unhappy with the implementation here (querying the DB directly, getting the ID from the URL is hacky, checking which page we're on via backtrace seems hacky, and there's only one place where we try to re-initialize the order currency) but wanted to get your 👀 on this to see what you think of the overall approach?

It fixes the issue on both the order received page as well as viewing the order details under the My Account &rarr; Orders page.

But it all just feels a bit hacky. Not really sure what else to do though. We basically need the order currency during the `wc_get_price_decimal_separator` hook, and we haven't called any function to set up the order currency by that point. And AFAICT there's no way to get the order currency except via direct query to the DB.

There's some more context in my comments on the issue #8232 but this is the gist of it.

Would love to pick your brain on this a bit.

**Comment by jessepearson** [comment]
I feel like what you're doing is similar to what `init_order_currency_from_query_vars` is doing, I am just not sure if that approach can be used here. Maybe if `$arg` is `null` when `init_order_currency` is called you can see if `$wp->query_vars` has the order number to reduce the amount of code.

I do agree that it seems hacky to get the order number from the URL, but we have to do that when it's not available, unfortunately. 

You mentioned that the `wc_get_order` call within `init_order_currency` is creating a cyclic call. Is it possible to remove actions/filters within `init_order_currency` before the call to `wc_get_order` and then add them back right after? There are [instances within the `WooCommerceSubscriptions` compatibility class](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/multi-currency/Compatibility/WooCommerceSubscriptions.php#L200) where we had to set a property in the class to make sure redundancies didn't happen, as well.

Dealing with these issues isn't easy because you can spend a couple days trying to solve the issue with different approaches and end up with like 4 lines of code to show for it.

**Comment by reykjalin** [comment]
> You mentioned that the `wc_get_order` call within `init_order_currency` is creating a cyclic call. Is it possible to remove actions/filters within `init_order_currency` before the call to `wc_get_order` and then add them back right after?

That was something I hadn't considered! I looked into this some more and you can take a look at the current approach. I think it's much better than the previous fix. Thanks for pointing me to this, I think that really helped improve the fix!

**Review by cesarcosta99 — APPROVED** [review]
Changes look good and test well! No deprecated warning was noted on PHP 8.1 :shipit:

File: includes/multi-currency/FrontendCurrencies.php [hunk 1]
```diff
@@ -270,7 +276,16 @@ public function init_order_currency( $arg ) {
...
+		// We remove the filters here becuase 'wc_get_order' triggers the 'wc_get_price_decimal_separator' filter.
+		remove_filter( 'wc_get_price_decimal_separator', [ $this, 'get_price_decimal_separator' ], 900 );
+		remove_filter( 'wc_get_price_decimal_separator', [ $this, 'init_order_currency_from_query_vars' ], 900 );
 		$order = ! $arg instanceof WC_Order ? wc_get_order( $arg ) : $arg;
+		// Note: it's important that 'init_order_currency_from_query_vars' is called before
+		// 'get_price_decimal_separator' because the order currency is often required to
+		// determine the decimal separator. That's why the priority on 'init_order_currency_from_query_vars'
+		// is explicity lower than the priority of 'get_price_decimal_separator'.
+		add_filter( 'wc_get_price_decimal_separator', [ $this, 'init_order_currency_from_query_vars' ], 900 );
+		add_filter( 'wc_get_price_decimal_separator', [ $this, 'get_price_decimal_separator' ], 900 );
```
Comment (cesarcosta99) [inline includes/multi-currency/FrontendCurrencies.php]: Just noticed this after approving, but according to your notes, the priority on `get_price_decimal_separator` should be `901` here as well.

File: includes/multi-currency/FrontendCurrencies.php [hunk 2]
```diff
@@ -270,7 +276,16 @@ public function init_order_currency( $arg ) {
 			return $arg;
 		}
 
+		// We remove the filters here becuase 'wc_get_order' triggers the 'wc_get_price_decimal_separator' filter.
+		remove_filter( 'wc_get_price_decimal_separator', [ $this, 'get_price_decimal_separator' ], 900 );
+		remove_filter( 'wc_get_price_decimal_separator', [ $this, 'init_order_currency_from_query_vars' ], 900 );
```
Comment (cesarcosta99) [inline includes/multi-currency/FrontendCurrencies.php]: Do we have to worry about the priority being the same here too? I don't think so but thought of asking to be sure 🙂

**Review by cesarcosta99 — CHANGES_REQUESTED** [review]

File: includes/multi-currency/FrontendCurrencies.php (hunk 2 above)
Comment (reykjalin) [reply includes/multi-currency/FrontendCurrencies.php]: yes, we do have to worry about the priority being the same, thank you for catching this! Completely missed it 😅

File: includes/multi-currency/FrontendCurrencies.php (hunk 1 above)
Comment (reykjalin) [reply includes/multi-currency/FrontendCurrencies.php]: Fixed in https://github.com/Automattic/woocommerce-payments/pull/8387/commits/3f93b1fdc324bfd7a1029265635e63c7afd5f42b

File: includes/multi-currency/FrontendCurrencies.php (hunk 2 above)
Comment (reykjalin) [reply includes/multi-currency/FrontendCurrencies.php]: Fixed in https://github.com/Automattic/woocommerce-payments/pull/8387/commits/3f93b1fdc324bfd7a1029265635e63c7afd5f42b

**Review by cesarcosta99 — APPROVED** [review]
LGTM 💯

**Pull request merged**